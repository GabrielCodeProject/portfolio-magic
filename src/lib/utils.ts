import { type ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function debounce<T extends (...args: unknown[]) => unknown>(
  func: T,
  wait: number,
  immediate?: boolean
): T {
  let timeout: NodeJS.Timeout | null = null;

  return ((...args: Parameters<T>) => {
    const later = () => {
      timeout = null;
      if (!immediate) func(...args);
    };

    const callNow = immediate && !timeout;

    if (timeout) clearTimeout(timeout);
    timeout = setTimeout(later, wait);

    if (callNow) func(...args);
  }) as T;
}

export function throttle<T extends (...args: unknown[]) => unknown>(
  func: T,
  limit: number
): T {
  let inThrottle: boolean;

  return ((...args: Parameters<T>) => {
    if (!inThrottle) {
      func(...args);
      inThrottle = true;
      setTimeout(() => (inThrottle = false), limit);
    }
  }) as T;
}

export function lerp(start: number, end: number, t: number): number {
  return start * (1 - t) + end * t;
}

export function clamp(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}

export function randomInRange(min: number, max: number): number {
  return Math.random() * (max - min) + min;
}

export function formatDate(date: Date): string {
  return new Intl.DateTimeFormat('en-US', {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  }).format(date);
}

export function slugify(text: string): string {
  return text
    .toLowerCase()
    .replace(/[^\w ]+/g, '')
    .replace(/ +/g, '-');
}
