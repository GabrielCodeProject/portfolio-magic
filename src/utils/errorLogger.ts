'use client';

// Enhanced client-side error logging utility for 3D components and fallbacks

export interface ErrorLogEntry {
  id: string;
  timestamp: string;
  component: string;
  componentType: '3D' | 'fallback' | 'boundary' | 'performance';
  severity: 'low' | 'medium' | 'high' | 'critical';
  error: {
    message: string;
    stack?: string;
    name: string;
  };
  context?: {
    userAgent?: string;
    url?: string;
    viewport?: { width: number; height: number };
    theme?: string;
    performanceData?: any;
    componentStack?: string;
  };
  sessionId: string;
}

export interface ErrorLoggerConfig {
  maxLogs: number;
  enableConsoleLogging: boolean;
  enableLocalStorage: boolean;
  logLevels: ('low' | 'medium' | 'high' | 'critical')[];
}

class ClientErrorLogger {
  private config: ErrorLoggerConfig;
  private sessionId: string;
  private storageKey = 'three-component-error-logs';

  constructor(config: Partial<ErrorLoggerConfig> = {}) {
    this.config = {
      maxLogs: 100,
      enableConsoleLogging: true,
      enableLocalStorage: true,
      logLevels: ['low', 'medium', 'high', 'critical'],
      ...config,
    };

    // Generate unique session ID
    this.sessionId = this.generateSessionId();
  }

  private generateSessionId(): string {
    return `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  private generateLogId(): string {
    return `log_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  private getContextInfo(): ErrorLogEntry['context'] {
    if (typeof window === 'undefined') return {};

    return {
      userAgent: navigator.userAgent,
      url: window.location.href,
      viewport: {
        width: window.innerWidth,
        height: window.innerHeight,
      },
      theme: document.documentElement.getAttribute('data-theme') || 'unknown',
    };
  }

  logError(
    error: Error,
    componentName: string,
    componentType: ErrorLogEntry['componentType'] = '3D',
    severity: ErrorLogEntry['severity'] = 'medium',
    additionalContext?: Partial<ErrorLogEntry['context']>
  ): void {
    // Check if this severity level should be logged
    if (!this.config.logLevels.includes(severity)) {
      return;
    }

    const logEntry: ErrorLogEntry = {
      id: this.generateLogId(),
      timestamp: new Date().toISOString(),
      component: componentName,
      componentType,
      severity,
      error: {
        message: error.message,
        stack: error.stack,
        name: error.name,
      },
      context: {
        ...this.getContextInfo(),
        ...additionalContext,
      },
      sessionId: this.sessionId,
    };

    // Console logging
    if (this.config.enableConsoleLogging) {
      this.logToConsole(logEntry);
    }

    // localStorage logging
    if (this.config.enableLocalStorage) {
      this.logToStorage(logEntry);
    }
  }

  private logToConsole(logEntry: ErrorLogEntry): void {
    const prefix = `[${logEntry.componentType.toUpperCase()}][${logEntry.severity.toUpperCase()}][${logEntry.component}]`;
    const message = `${prefix} ${logEntry.error.message}`;

    switch (logEntry.severity) {
      case 'critical':
        console.error(message, logEntry);
        break;
      case 'high':
        console.error(message, logEntry);
        break;
      case 'medium':
        console.warn(message, logEntry);
        break;
      case 'low':
        console.info(message, logEntry);
        break;
      default:
        console.log(message, logEntry);
    }
  }

  private logToStorage(logEntry: ErrorLogEntry): void {
    if (typeof window === 'undefined') return;

    try {
      const existingLogs = this.getLogs();
      existingLogs.push(logEntry);

      // Keep only the most recent logs
      if (existingLogs.length > this.config.maxLogs) {
        existingLogs.splice(0, existingLogs.length - this.config.maxLogs);
      }

      localStorage.setItem(this.storageKey, JSON.stringify(existingLogs));
    } catch (storageError) {
      console.error('[ErrorLogger] Failed to save to localStorage:', storageError);
    }
  }

  getLogs(): ErrorLogEntry[] {
    if (typeof window === 'undefined') return [];

    try {
      const logs = localStorage.getItem(this.storageKey);
      return logs ? JSON.parse(logs) : [];
    } catch (parseError) {
      console.error('[ErrorLogger] Failed to parse logs from localStorage:', parseError);
      return [];
    }
  }

  getLogsByComponent(componentName: string): ErrorLogEntry[] {
    return this.getLogs().filter(log => log.component === componentName);
  }

  getLogsBySeverity(severity: ErrorLogEntry['severity']): ErrorLogEntry[] {
    return this.getLogs().filter(log => log.severity === severity);
  }

  getLogsForSession(sessionId?: string): ErrorLogEntry[] {
    const targetSessionId = sessionId || this.sessionId;
    return this.getLogs().filter(log => log.sessionId === targetSessionId);
  }

  clearLogs(): void {
    if (typeof window === 'undefined') return;

    try {
      localStorage.removeItem(this.storageKey);
    } catch (error) {
      console.error('[ErrorLogger] Failed to clear logs:', error);
    }
  }

  exportLogs(): string {
    return JSON.stringify(this.getLogs(), null, 2);
  }

  // Helper method for React Error Boundaries
  logReactError(
    error: Error,
    errorInfo: React.ErrorInfo,
    componentName: string,
    severity: ErrorLogEntry['severity'] = 'high'
  ): void {
    this.logError(error, componentName, 'boundary', severity, {
      componentStack: errorInfo.componentStack || undefined,
    });
  }

  // Helper method for performance-related errors
  logPerformanceError(
    error: Error,
    componentName: string,
    performanceData: any,
    severity: ErrorLogEntry['severity'] = 'medium'
  ): void {
    this.logError(error, componentName, 'performance', severity, {
      performanceData,
    });
  }

  // Helper method for fallback component issues
  logFallbackError(
    error: Error,
    componentName: string,
    severity: ErrorLogEntry['severity'] = 'low'
  ): void {
    this.logError(error, componentName, 'fallback', severity);
  }
}

// Create singleton instance
export const errorLogger = new ClientErrorLogger();

// Export types and utilities
export { ClientErrorLogger };
export default errorLogger;