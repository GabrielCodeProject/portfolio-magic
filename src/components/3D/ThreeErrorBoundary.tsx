'use client';

import React, { Component, ReactNode } from 'react';

import { errorLogger } from '@/utils/errorLogger';

interface Props {
  children: ReactNode;
  fallback?: ReactNode;
  componentName?: string;
  severity?: 'low' | 'medium' | 'high' | 'critical';
}

interface State {
  hasError: boolean;
  error?: Error;
  errorId?: string;
}

class ThreeErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error: Error): State {
    return { hasError: true, error };
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    const { componentName = 'Unknown3DComponent', severity = 'high' } = this.props;
    
    // Enhanced client-side error logging with better component identification
    try {
      errorLogger.logReactError(error, errorInfo, componentName, severity);
      
      // Store error ID for potential debugging
      this.setState({ 
        errorId: `${componentName}_${Date.now()}` 
      });

      // Additional context logging for 3D-specific issues
      const is3DError = this.detect3DSpecificError(error);
      if (is3DError) {
        console.group(`ðŸŽ® 3D Component Error: ${componentName}`);
        console.error('Error Details:', error);
        console.error('Component Stack:', errorInfo.componentStack);
        console.error('Possible causes:', this.suggest3DErrorCauses(error));
        console.groupEnd();
      }

    } catch (loggingError) {
      // Fallback logging if our enhanced logger fails
      console.error('Enhanced error logging failed, falling back to basic logging:', loggingError);
      console.error(`ThreeErrorBoundary caught error in ${componentName}:`, error, errorInfo);
    }
  }

  private detect3DSpecificError(error: Error): boolean {
    const errorMessage = error.message.toLowerCase();
    const stack = error.stack?.toLowerCase() || '';
    
    const threejsKeywords = [
      'webgl', 'three', 'shader', 'buffer', 'geometry', 'material',
      'texture', 'canvas', 'context', 'renderer', 'scene', 'camera'
    ];
    
    return threejsKeywords.some(keyword => 
      errorMessage.includes(keyword) || stack.includes(keyword)
    );
  }

  private suggest3DErrorCauses(error: Error): string[] {
    const suggestions: string[] = [];
    const errorMessage = error.message.toLowerCase();
    
    if (errorMessage.includes('webgl')) {
      suggestions.push('WebGL not supported or context lost');
      suggestions.push('GPU drivers may need updating');
    }
    
    if (errorMessage.includes('shader')) {
      suggestions.push('Shader compilation failed');
      suggestions.push('Graphics card may not support required features');
    }
    
    if (errorMessage.includes('buffer') || errorMessage.includes('memory')) {
      suggestions.push('Insufficient GPU memory');
      suggestions.push('Too many 3D objects loaded simultaneously');
    }
    
    if (errorMessage.includes('texture')) {
      suggestions.push('Texture loading failed');
      suggestions.push('Image format not supported');
    }
    
    if (suggestions.length === 0) {
      suggestions.push('Generic 3D rendering error');
      suggestions.push('Check browser console for more details');
    }
    
    return suggestions;
  }

  render() {
    if (this.state.hasError) {
      // Return null for Canvas contexts to avoid R3F namespace errors
      // HTML elements cannot be rendered inside React Three Fiber Canvas
      return this.props.fallback || null;
    }

    return this.props.children;
  }
}

export default ThreeErrorBoundary;