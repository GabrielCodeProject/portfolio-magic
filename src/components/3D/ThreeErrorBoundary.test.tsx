import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import ThreeErrorBoundary from './ThreeErrorBoundary';
import { errorLogger } from '@/utils/errorLogger';

// Mock the errorLogger
jest.mock('@/utils/errorLogger', () => ({
  errorLogger: {
    logReactError: jest.fn(),
  },
}));

// Component that throws an error on purpose for testing
const ThrowError: React.FC<{ shouldThrow?: boolean; errorMessage?: string }> = ({ 
  shouldThrow = true, 
  errorMessage = 'Test error' 
}) => {
  if (shouldThrow) {
    throw new Error(errorMessage);
  }
  return <div data-testid="success-component">Component rendered successfully</div>;
};

// Component that renders successfully
const SuccessComponent: React.FC = () => (
  <div data-testid="success-component">Component rendered successfully</div>
);

// Mock console methods to test logging
const mockConsoleError = jest.spyOn(console, 'error').mockImplementation(() => {});
const mockConsoleGroup = jest.spyOn(console, 'group').mockImplementation(() => {});
const mockConsoleGroupEnd = jest.spyOn(console, 'groupEnd').mockImplementation(() => {});

describe('ThreeErrorBoundary', () => {
  const originalEnv = process.env.NODE_ENV;

  beforeEach(() => {
    jest.clearAllMocks();
    mockConsoleError.mockClear();
    mockConsoleGroup.mockClear();
    mockConsoleGroupEnd.mockClear();
  });

  afterEach(() => {
    process.env.NODE_ENV = originalEnv;
  });

  describe('Normal Operation', () => {
    it('should render children when no error occurs', () => {
      render(
        <ThreeErrorBoundary>
          <SuccessComponent />
        </ThreeErrorBoundary>
      );

      expect(screen.getByTestId('success-component')).toBeInTheDocument();
      expect(screen.getByText('Component rendered successfully')).toBeInTheDocument();
    });

    it('should render multiple children when no error occurs', () => {
      render(
        <ThreeErrorBoundary>
          <div data-testid="child-1">Child 1</div>
          <div data-testid="child-2">Child 2</div>
        </ThreeErrorBoundary>
      );

      expect(screen.getByTestId('child-1')).toBeInTheDocument();
      expect(screen.getByTestId('child-2')).toBeInTheDocument();
    });

    it('should pass through props to children', () => {
      const TestComponent: React.FC<{ testProp: string }> = ({ testProp }) => (
        <div data-testid="test-component">{testProp}</div>
      );

      render(
        <ThreeErrorBoundary>
          <TestComponent testProp="test value" />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('test value')).toBeInTheDocument();
    });
  });

  describe('Error Catching', () => {
    it('should catch errors and display default fallback UI', () => {
      render(
        <ThreeErrorBoundary>
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
      expect(screen.getByText('⚠️')).toBeInTheDocument();
    });

    it('should catch errors with custom error messages', () => {
      render(
        <ThreeErrorBoundary>
          <ThrowError errorMessage="Custom WebGL error" />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
    });

    it('should update component state when error occurs', () => {
      const TestBoundary = () => {
        const [hasError, setHasError] = React.useState(false);
        
        React.useEffect(() => {
          const boundary = document.querySelector('.error-boundary');
          if (boundary?.textContent?.includes('3D component failed to load')) {
            setHasError(true);
          }
        }, []);

        return (
          <div>
            <div className="error-boundary">
              <ThreeErrorBoundary>
                <ThrowError />
              </ThreeErrorBoundary>
            </div>
            <div data-testid="error-state">{hasError ? 'error-caught' : 'no-error'}</div>
          </div>
        );
      };

      render(<TestBoundary />);
      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
    });
  });

  describe('Custom Fallback UI', () => {
    it('should render custom fallback when provided', () => {
      const customFallback = <div data-testid="custom-fallback">Custom fallback UI</div>;

      render(
        <ThreeErrorBoundary fallback={customFallback}>
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByTestId('custom-fallback')).toBeInTheDocument();
      expect(screen.getByText('Custom fallback UI')).toBeInTheDocument();
      expect(screen.queryByText('3D component failed to load')).not.toBeInTheDocument();
    });

    it('should render complex custom fallback components', () => {
      const CustomFallback = () => (
        <div data-testid="complex-fallback">
          <h2>Something went wrong</h2>
          <p>Please try refreshing the page</p>
          <button>Retry</button>
        </div>
      );

      render(
        <ThreeErrorBoundary fallback={<CustomFallback />}>
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByTestId('complex-fallback')).toBeInTheDocument();
      expect(screen.getByText('Something went wrong')).toBeInTheDocument();
      expect(screen.getByText('Please try refreshing the page')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Retry' })).toBeInTheDocument();
    });
  });

  describe('Error Logging Integration', () => {
    it('should call errorLogger.logReactError when error occurs', () => {
      const mockLogReactError = errorLogger.logReactError as jest.Mock;

      render(
        <ThreeErrorBoundary componentName="TestComponent" severity="high">
          <ThrowError errorMessage="Test logging error" />
        </ThreeErrorBoundary>
      );

      expect(mockLogReactError).toHaveBeenCalledWith(
        expect.objectContaining({
          message: 'Test logging error',
          name: 'Error'
        }),
        expect.objectContaining({
          componentStack: expect.any(String)
        }),
        'TestComponent',
        'high'
      );
    });

    it('should use default component name when not provided', () => {
      const mockLogReactError = errorLogger.logReactError as jest.Mock;

      render(
        <ThreeErrorBoundary>
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(mockLogReactError).toHaveBeenCalledWith(
        expect.any(Error),
        expect.any(Object),
        'Unknown3DComponent',
        'high'
      );
    });

    it('should use default severity when not provided', () => {
      const mockLogReactError = errorLogger.logReactError as jest.Mock;

      render(
        <ThreeErrorBoundary componentName="TestComponent">
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(mockLogReactError).toHaveBeenCalledWith(
        expect.any(Error),
        expect.any(Object),
        'TestComponent',
        'high'
      );
    });

    it('should handle errorLogger failures gracefully', () => {
      const mockLogReactError = errorLogger.logReactError as jest.Mock;
      mockLogReactError.mockImplementationOnce(() => {
        throw new Error('Logger failed');
      });

      render(
        <ThreeErrorBoundary componentName="TestComponent">
          <ThrowError />
        </ThreeErrorBoundary>
      );

      // Should still render fallback UI even if logging fails
      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
      expect(mockConsoleError).toHaveBeenCalledWith(
        'Enhanced error logging failed, falling back to basic logging:',
        expect.any(Error)
      );
    });
  });

  describe('3D-Specific Error Detection', () => {
    it('should detect WebGL errors', () => {
      render(
        <ThreeErrorBoundary componentName="WebGLComponent">
          <ThrowError errorMessage="WebGL context lost" />
        </ThreeErrorBoundary>
      );

      expect(mockConsoleGroup).toHaveBeenCalledWith('🎮 3D Component Error: WebGLComponent');
      expect(mockConsoleError).toHaveBeenCalledWith('Possible causes:', expect.arrayContaining([
        'WebGL not supported or context lost',
        'GPU drivers may need updating'
      ]));
      expect(mockConsoleGroupEnd).toHaveBeenCalled();
    });

    it('should detect Three.js errors', () => {
      render(
        <ThreeErrorBoundary componentName="ThreeComponent">
          <ThrowError errorMessage="THREE.Scene is not defined" />
        </ThreeErrorBoundary>
      );

      expect(mockConsoleGroup).toHaveBeenCalledWith('🎮 3D Component Error: ThreeComponent');
      expect(mockConsoleGroupEnd).toHaveBeenCalled();
    });

    it('should detect shader errors', () => {
      render(
        <ThreeErrorBoundary componentName="ShaderComponent">
          <ThrowError errorMessage="Shader compilation failed" />
        </ThreeErrorBoundary>
      );

      expect(mockConsoleError).toHaveBeenCalledWith('Possible causes:', expect.arrayContaining([
        'Shader compilation failed',
        'Graphics card may not support required features'
      ]));
    });

    it('should detect buffer/memory errors', () => {
      render(
        <ThreeErrorBoundary componentName="BufferComponent">
          <ThrowError errorMessage="Buffer allocation failed" />
        </ThreeErrorBoundary>
      );

      expect(mockConsoleError).toHaveBeenCalledWith('Possible causes:', expect.arrayContaining([
        'Insufficient GPU memory',
        'Too many 3D objects loaded simultaneously'
      ]));
    });

    it('should detect texture errors', () => {
      render(
        <ThreeErrorBoundary componentName="TextureComponent">
          <ThrowError errorMessage="Texture loading failed" />
        </ThreeErrorBoundary>
      );

      expect(mockConsoleError).toHaveBeenCalledWith('Possible causes:', expect.arrayContaining([
        'Texture loading failed',
        'Image format not supported'
      ]));
    });

    it('should provide generic suggestions for non-3D errors', () => {      
      // Test with an error that definitely has no 3D keywords
      const mockDetect = jest.spyOn(ThreeErrorBoundary.prototype as any, 'detect3DSpecificError');
      mockDetect.mockReturnValue(false); // Force non-3D detection
      
      render(
        <ThreeErrorBoundary componentName="GenericComponent">
          <ThrowError errorMessage="Something went wrong in the component" />
        </ThreeErrorBoundary>
      );

      // Should not trigger 3D-specific logging for non-3D errors
      expect(mockConsoleGroup).not.toHaveBeenCalledWith('🎮 3D Component Error: GenericComponent');
      
      mockDetect.mockRestore();
    });
  });

  describe('Development vs Production Behavior', () => {
    it('should show error ID in development mode', () => {
      Object.defineProperty(process.env, 'NODE_ENV', { value: 'development', writable: true });

      render(
        <ThreeErrorBoundary componentName="DevComponent">
          <ThrowError />
        </ThreeErrorBoundary>
      );

      // Should show error ID pattern
      expect(screen.getByText(/Error ID: DevComponent_\d+/)).toBeInTheDocument();
    });

    it('should not show error ID in production mode', () => {
      Object.defineProperty(process.env, 'NODE_ENV', { value: 'production', writable: true });

      render(
        <ThreeErrorBoundary componentName="ProdComponent">
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.queryByText(/Error ID:/)).not.toBeInTheDocument();
    });

    it('should handle undefined NODE_ENV', () => {
      delete (process.env as any).NODE_ENV;

      render(
        <ThreeErrorBoundary componentName="UndefComponent">
          <ThrowError />
        </ThreeErrorBoundary>
      );

      // Should render fallback without error
      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
    });
  });

  describe('Error ID Generation', () => {
    it('should generate unique error IDs', () => {
      Object.defineProperty(process.env, 'NODE_ENV', { value: 'development', writable: true });

      // Create first error boundary instance
      const { container: container1 } = render(
        <div data-testid="boundary-1">
          <ThreeErrorBoundary componentName="Component1">
            <ThrowError />
          </ThreeErrorBoundary>
        </div>
      );

      const firstErrorElement = container1.querySelector('[data-testid="boundary-1"]')?.textContent;
      const firstErrorIdMatch = firstErrorElement?.match(/Error ID: (Component1_\d+)/);
      const firstErrorId = firstErrorIdMatch?.[1];

      // Create second error boundary instance  
      const { container: container2 } = render(
        <div data-testid="boundary-2">
          <ThreeErrorBoundary componentName="Component2">
            <ThrowError />
          </ThreeErrorBoundary>
        </div>
      );

      const secondErrorElement = container2.querySelector('[data-testid="boundary-2"]')?.textContent;
      const secondErrorIdMatch = secondErrorElement?.match(/Error ID: (Component2_\d+)/);
      const secondErrorId = secondErrorIdMatch?.[1];

      expect(firstErrorId).toBeTruthy();
      expect(secondErrorId).toBeTruthy();
      expect(firstErrorId).not.toBe(secondErrorId);
    });

    it('should include component name in error ID', () => {
      Object.defineProperty(process.env, 'NODE_ENV', { value: 'development', writable: true });

      render(
        <ThreeErrorBoundary componentName="MySpecialComponent">
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText(/Error ID: MySpecialComponent_\d+/)).toBeInTheDocument();
    });
  });

  describe('Error State Management', () => {
    it('should maintain error state after initial error', () => {
      const { rerender } = render(
        <ThreeErrorBoundary>
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();

      // Rerender with successful component - should still show error
      rerender(
        <ThreeErrorBoundary>
          <SuccessComponent />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
      expect(screen.queryByTestId('success-component')).not.toBeInTheDocument();
    });

    it('should reset error state with new boundary instance', () => {
      const { unmount } = render(
        <ThreeErrorBoundary>
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();

      unmount();

      render(
        <ThreeErrorBoundary>
          <SuccessComponent />
        </ThreeErrorBoundary>
      );

      expect(screen.getByTestId('success-component')).toBeInTheDocument();
      expect(screen.queryByText('3D component failed to load')).not.toBeInTheDocument();
    });
  });

  describe('Component Lifecycle', () => {
    it('should initialize with no error state', () => {
      const { container } = render(
        <ThreeErrorBoundary>
          <div data-testid="initial-child">Initial content</div>
        </ThreeErrorBoundary>
      );

      expect(screen.getByTestId('initial-child')).toBeInTheDocument();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should handle componentDidCatch lifecycle', () => {
      const mockLogReactError = errorLogger.logReactError as jest.Mock;

      render(
        <ThreeErrorBoundary componentName="LifecycleTest">
          <ThrowError errorMessage="Lifecycle test error" />
        </ThreeErrorBoundary>
      );

      expect(mockLogReactError).toHaveBeenCalledTimes(1);
      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
    });

    it('should handle unmounting gracefully', () => {
      const { unmount } = render(
        <ThreeErrorBoundary>
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();

      // Should unmount without throwing
      expect(() => unmount()).not.toThrow();
    });
  });

  describe('Edge Cases', () => {
    it('should handle errors with no message', () => {
      const ErrorWithNoMessage = () => {
        const error = new Error();
        error.message = '';
        throw error;
      };

      render(
        <ThreeErrorBoundary>
          <ErrorWithNoMessage />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
    });

    it('should handle errors with no stack trace', () => {
      const ErrorWithNoStack = () => {
        const error = new Error('No stack error');
        error.stack = undefined;
        throw error;
      };

      render(
        <ThreeErrorBoundary>
          <ErrorWithNoStack />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
    });

    it('should handle null children', () => {
      render(
        <ThreeErrorBoundary>
          {null}
        </ThreeErrorBoundary>
      );

      // Should render nothing but not crash
      expect(document.body).toBeInTheDocument();
    });

    it('should handle undefined children', () => {
      render(
        <ThreeErrorBoundary>
          {undefined}
        </ThreeErrorBoundary>
      );

      // Should render nothing but not crash
      expect(document.body).toBeInTheDocument();
    });

    it('should handle empty children array', () => {
      render(
        <ThreeErrorBoundary>
          {[]}
        </ThreeErrorBoundary>
      );

      // Should render nothing but not crash
      expect(document.body).toBeInTheDocument();
    });
  });

  describe('Error Recovery', () => {
    it('should not recover from errors automatically', () => {
      const { rerender } = render(
        <ThreeErrorBoundary>
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();

      // Force re-render - should still show error
      rerender(
        <ThreeErrorBoundary>
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
    });

    it('should handle prop changes after error', () => {
      const { rerender } = render(
        <ThreeErrorBoundary componentName="Initial">
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();

      // Change props - should show new fallback when fallback prop changes
      rerender(
        <ThreeErrorBoundary 
          componentName="Updated" 
          fallback={<div data-testid="updated-fallback">Updated fallback</div>}
        >
          <ThrowError />
        </ThreeErrorBoundary>
      );

      expect(screen.getByTestId('updated-fallback')).toBeInTheDocument();
      expect(screen.getByText('Updated fallback')).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it('should have accessible error message structure', () => {
      render(
        <ThreeErrorBoundary>
          <ThrowError />
        </ThreeErrorBoundary>
      );

      const errorContainer = screen.getByText('3D component failed to load').closest('div');
      expect(errorContainer).toHaveClass('text-center');
      expect(screen.getByText('⚠️')).toBeInTheDocument();
    });

    it('should maintain focus management during error display', () => {
      render(
        <div>
          <button data-testid="before-button">Before</button>
          <ThreeErrorBoundary>
            <ThrowError />
          </ThreeErrorBoundary>
          <button data-testid="after-button">After</button>
        </div>
      );

      expect(screen.getByTestId('before-button')).toBeInTheDocument();
      expect(screen.getByText('3D component failed to load')).toBeInTheDocument();
      expect(screen.getByTestId('after-button')).toBeInTheDocument();
    });
  });
});