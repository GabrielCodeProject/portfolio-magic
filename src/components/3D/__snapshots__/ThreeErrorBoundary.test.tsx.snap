// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ThreeErrorBoundary Component Lifecycle should initialize with no error state 1`] = `
<div
  data-testid="initial-child"
>
  Initial content
</div>
`;
